### ISSUE TITLE
Improving the Intent Classification Algorithm

### ISSUE THREAD

**Issue by AdnanKhalid1** [RasaHQ/rasa#4893]
**Description of Problem**:
Currently, the pre-trained word embeddings from Spacy are implemented such that RASA takes average of all the word vectors in a sentence and matches that with the average of word vectors in labelled Intents.  This may add some distortion/noise to the important words, specially if sentences are long and important word vectors may get washed away by unimportant ones, and this can reduce the intent classification accuracy.

**Overview of the Solution**:
It will thus be beneficial to somehow pre-process the training data containing the labelled intents, such that it automatically shortlist only the "important" words in the training sentences, and then we take the average of vectors for these shortlisted words only.  

To do this, we can first group together all training sentences for a specific intent as one document, and then for each word in the document we calculate an anomaly score of its vector using Isolation Forest from sklearn.  We then just exclude the top 50 anomolous words whose vectors are very different from the rest of the words in the document.  We are then left with words whose vectors are more or less similar or the words whose vectors more precisely define the intent.  We then just take average of these vectors after eliminating noise from the document.  

To match this with the sentence entered by a user, we however shouldn't shortlist words based on anomaly score as there may be few words in that sentence and it may be the anomolous words in that sentence that may be more predictive of the intent.  This is one of the reason I suggested anomaly score instead of TF-IDF for the labelled intents documents as well.  

**Definition of Done**:
Optional parameters added to the processing pipeline that lets users eliminate top 50% anomolous words from the document containing list of sentences for a specific intent.

Users can then test accuracies with and without this feature in their data.

(p.s Can somebody try this approach and show their results in comments below?)

**Comment by sara-tagger** [comment]
Thanks for submitting this feature request 🚀@ArjaanBuijk will get back to you about it soon!✨

**Comment by dakshvar22** [comment]
@AdnanKhalid1 You are absolutely right that averaging word vectors in the spacy pipeline is detrimental. We are working on a new model architecture for NLU currently which does not average out the vectors but essentially uses a sequential model to preserve information across words. 
We plan to release it sometime in January.

**Comment by code4passion** [comment]
Any updates on this? I though release was in jan

**Comment by dakshvar22** [comment]
@code4passion Yes, it should be out with the next minor release next week.
